amethyst_tiles/src/morton/mod.rs

```diff
@@ -0,0 +1,231 @@
+}
+
+#[inline]
+pub fn morton_decode_intr_3d(morton: u32) -> (u32, u32, u32) {
+    use bitintr::Pext;
+
+    (
+        morton.pext(0x0924_9249),
+        morton.pext(0x1249_2492),
+        morton.pext(0x2492_4924),
+    )
+}
+
+#[inline]
+#[cfg(target_feature = "bmi2")]
+pub fn encode(x: u32, y: u32, z: u32) -> u32 {
+    morton_encode_intr_3d(x, y, z)
+}
+
+#[inline]
+#[cfg(target_feature = "bmi2")]
+pub fn decode(morton: u32) -> (u32, u32, u32) {
+    morton_decode_intr_3d(morton)
+}
+
+/// 3D Morton (Z-Order) encoding implementation.
+/// This implementation uses the `bmi2` CPU intrinsic if it is available via the `bitintr` crate. If this instruction
+/// set is not available, it falls back on simpler computation methods. Using these CPU instruction optimizations requires
+/// `RUSTFLAGS=-C target-feature=+bmi2`. If this target feature is not provided, a LUT (Look Up Table) implementation
+/// of Morton encoding is used, considered extremely fast but still slightly slower than BMI2 intrinsics.
+#[derive(Default, Clone)]
+pub struct Encoder;
+impl CoordinateEncoder for Encoder {
+    fn from_dimensions(_: u32, _: u32, _: u32) -> Self {
+        Self {}
+    }
+
+    #[inline]
+    fn encode(&self, x: u32, y: u32, z: u32) -> Option<u32> {
+        Some(encode(x, y, z))
+    }
+    #[inline]
+    fn decode(&self, morton: u32) -> Option<(u32, u32, u32)> {
+        Some(decode(morton))
+    }
+}
+
+/// 3D Morton (Z-Order) encoding implementation.
+/// This implementation uses the `bmi2` CPU intrinsic if it is available via the `bitintr` crate. If this instruction
+/// set is not available, it falls back on simpler computation methods. Using these CPU instruction optimizations requires
+/// `RUSTFLAGS=-C target-feature=+bmi2`. If this target feature is not provided, a LUT (Look Up Table) implementation
+/// of Morton encoding is used, considered extremely fast but still slightly slower than BMI2 intrinsics.
+///
+/// This implementation only performs 2D morton encoding on any given Z-level, while providing Z-levels ia a standard
+/// flat-array multiplicative manner. This means that each Z-level is contiguous in memory, but its inner coordinates
+/// are still Z-order encoded for some spatial locality.
+#[derive(Default, Clone)]
+pub struct Encoder2D {
+    dimensions: (u32, u32, u32),
+    len: u32,
+}
+impl CoordinateEncoder for Encoder2D {
+    fn from_dimensions(x: u32, y: u32, z: u32) -> Self {
+        //#[cfg(debug_assertions)]
+        //{
+        //    assert_eq!(x % 16, 0);
+        //    assert_eq!(x % 16, 0);
+        //}
+
+        Self {
+            dimensions: (x, y, z),
+            len: x * y,
+        }
+    }
+
+    #[inline]
+    fn encode(&self, x: u32, y: u32, z: u32) -> Option<u32> {
+        use bitintr::Pdep;
+
+        #[cfg(debug_assertions)]
+        {
+            let check = u32::max_value() / 3;
+            if x > check || y > check || z > check {
+                return None;
```

Comment:
I put this under a debug flag to assert that the requested coordinates encode to a valid morton value. In practice (release) all code should be written to prevent this. In debug, this is something you would like to build against. 

I've changed this to a panic in debug, but I am leaving it as a debug check.

The idea here is eventually (tm) this will support 64-bit mortons as well.

Reply:
seems good 👍